### PR TITLE
API: Add usage stats preview endpoint

### PR DESCRIPTION
**Pull request by Jguer**
**What this PR does / why we need it**:

Adds usage stats preview API endpoint. 

This patch has made it easier to test adding new usage statistics and it would also give a transparent way for users to see what usage statistics are being sent to the provider.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Todo after merge**:

- accompanying PR for docs

- Merge [Fix test broken](https://github.com/grafana/grafana-enterprise/pull/2641)

### CODE REVIEW

File: pkg/infra/usagestats/service/api.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+func (uss *UsageStats) getUsageReportPreview(ctx *models.ReqContext) response.Response {
+	if !uss.Cfg.ReportingEnabled {
+		return response.JSON(http.StatusUnauthorized, "Reporting is not enabled")
+	}
+
+	usageReport, err := uss.GetUsageReport(ctx.Req.Context())
+	if err != nil {
+		return response.Error(http.StatusInternalServerError, "failed to get summary for dashboard", err)
```
Comment (gamab) [inline pkg/infra/usagestats/service/api.go:27]:
```suggestion
		return response.Error(http.StatusInternalServerError, "failed to get usage report", err)
```

File: pkg/api/api.go [hunk 2]
```diff
@@ -70,6 +70,7 @@ func (hs *HTTPServer) registerRoutes() {
 	r.Get("/admin/orgs", authorizeInOrg(reqGrafanaAdmin, acmiddleware.UseGlobalOrg, orgsAccessEvaluator), hs.Index)
 	r.Get("/admin/orgs/edit/:id", authorizeInOrg(reqGrafanaAdmin, acmiddleware.UseGlobalOrg, orgsAccessEvaluator), hs.Index)
 	r.Get("/admin/stats", authorize(reqGrafanaAdmin, ac.EvalPermission(ac.ActionServerStatsRead)), hs.Index)
+	r.Get("/admin/usage-report-preview", authorize(reqGrafanaAdmin, ac.EvalPermission(ac.ActionServerStatsRead)), hs.Index)
```
Comment (gamab) [inline pkg/api/api.go:73]: Why are we registering this route, there don't seem to be any frontend associated?
image

**Review by gamab — APPROVED** [review]
LGTM, however, I'm not sure why we register the `/admin/usage-report-preview` route.

**Review by gamab — CHANGES_REQUESTED** [review]
Just to make sure we don't introduce any side effect, I'd like us to first check that we are not resetting any stats when hitting the preview endpoint, before merging, if you don't mind :slightly_smiling_face:

File: pkg/api/api.go (hunk 2 above)
Comment (Jguer) [reply pkg/api/api.go:73]: Indeed, we should not register it.

Context: Talked with Mitch about frontend, he thinks for now we can just expose an API endpoint

File: pkg/infra/usagestats/service/api.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+func (uss *UsageStats) registerAPIEndpoints() {
+	uss.RouteRegister.Group(rootUrl, func(subrouter routing.RouteRegister) {
+		subrouter.Get("/usage-report-preview", middleware.ReqGrafanaAdmin, routing.Wrap(uss.getUsageReportPreview))
+	})
+}
+
+func (uss *UsageStats) getUsageReportPreview(ctx *models.ReqContext) response.Response {
+	if !uss.Cfg.ReportingEnabled {
+		return response.JSON(http.StatusUnauthorized, "Reporting is not enabled")
```
Comment (DanCech) [inline pkg/infra/usagestats/service/api.go:22]: we might want to think about this endpoint still being available even if the actual sending is disabled, so users can review it to verify before enabling if they're concerned about privacy

Comment (Jguer) [reply pkg/infra/usagestats/service/api.go:22]: That's a good point, will implement

File: pkg/infra/usagestats/service/api.go (hunk 1 above)
Comment (DanCech) [reply pkg/infra/usagestats/service/api.go:23]: this still needs to be fixed

**Comment by DanCech** [comment]
> Just to make sure we don't introduce any side effect, I'd like us to first check that we are not resetting any stats when hitting the preview endpoint, before merging, if you don't mind 🙂

The same usage report is used for licensing purposes in enterprise, so this *should* be safe.

**Review by gamab — APPROVED** [review]

File: pkg/infra/usagestats/service/api.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	"github.com/grafana/grafana/pkg/api/response"
+	"github.com/grafana/grafana/pkg/api/routing"
+	"github.com/grafana/grafana/pkg/middleware"
+	"github.com/grafana/grafana/pkg/models"
+)
+
+const rootUrl = "/api/admin"
+
+func (uss *UsageStats) registerAPIEndpoints() {
+	uss.RouteRegister.Group(rootUrl, func(subrouter routing.RouteRegister) {
```
Comment (vtorosyan) [inline pkg/infra/usagestats/service/api.go:15]: I know I am a late on this, but I think we need to have FGAC permissions on this endpoint too. As we are already restricting the Admin Stats page and relevant endpoints.

Comment (vtorosyan) [reply pkg/infra/usagestats/service/api.go:15]: As we talked in Slack, I was confused with server stats and usage stats. As this is about usage stats, we don't have yet fine-grained permissions on it, so should be fine to leave it like this for the time being.

**Pull request merged**